File: typescript/tests/integration.live.mjs
/**
 * Live integration test (v0.2.0) — runs against a REAL backend.
 *
 * Usage:
 *   VIDEOFETCH_API_KEY=vf_live_sk_... VIDEOFETCH_BASE_URL=http://127.0.0.1:8301 \
 *     node tests/integration.live.mjs
 *
 * The key is read from the environment (or VIDEOFERCH_KEY_FILE / /tmp/vfkey.json)
 * and is NEVER written into the repository.
 *
 * Prereq: `npm run build` (this imports the built dist/).
 */
import { readFileSync } from "node:fs";
import {
  VideoFetch, RateLimitError, SignatureVerificationError,
  constructEvent, computeSignature, WEBHOOK_EVENTS,
} from "../dist/index.js";

const BASE = process.env.VIDEOFETCH_BASE_URL || "http://127.0.0.1:8301";
const PROBE_URL = "https://www.youtube.com/watch?v=dQw4w9WgXcQ";

function loadKey() {
  if (process.env.VIDEOFETCH_API_KEY) return process.env.VIDEOFETCH_API_KEY;
  const file = process.env.VIDEOFETCH_KEY_FILE || "/tmp/vfkey.json";
  try {
    return JSON.parse(readFileSync(file, "utf8")).api_key;
  } catch {
    return undefined;
  }
}

const pass = (m) => console.log(`  ok   ${m}`);
const info = (m) => console.log(`  ·    ${m}`);
let failures = 0;
function check(cond, label) {
  if (cond) pass(label);
  else { failures += 1; console.error(`  FAIL ${label}`); }
}

async function main() {
  const apiKey = loadKey();
  if (!apiKey) {
    console.error("No API key. Set VIDEOFETCH_API_KEY (or provide /tmp/vfkey.json).");
    process.exit(2);
  }
  console.log(`\n== TS SDK v0.2.0 live integration @ ${BASE} ==`);

  const client = new VideoFetch({ apiKey, baseUrl: BASE, timeoutMs: 15_000, maxRetries: 0 });

  // ── usage.get() ──────────────────────────────────────────────────────────
  const usage = await client.usage.get();
  info(`usage.get(): plan=${usage.plan} quota_gb=${usage.quota_gb} remaining_gb=${usage.remaining_gb} ` +
       `concurrency_limit=${usage.concurrency_limit} alert_level=${usage.alert_level}`);
  check(usage.quota_gb === 1.0, "usage.get() quota_gb === 1.0 (free tier)");
  check(usage.concurrency_limit === 5, "usage.get() concurrency_limit === 5");
  check(typeof usage.key_id === "string" && usage.key_id.length > 0, "usage.get() has key_id");
  check(typeof usage.month === "string" && /^\d{4}-\d{2}$/.test(usage.month), `usage.get() month=${usage.month}`);
  check(["ok", "warning", "critical", "exceeded"].includes(usage.alert_level), `usage.get() alert_level=${usage.alert_level}`);

  // ── usage.alerts() ───────────────────────────────────────────────────────
  const alerts = await client.usage.alerts();
  info(`usage.alerts(): thresholds=${JSON.stringify(alerts.thresholds)} topup=${JSON.stringify(alerts.topup_amounts)} ` +
       `level=${alerts.state.level} crossed=${JSON.stringify(alerts.state.crossed)}`);
  check(JSON.stringify(alerts.thresholds) === JSON.stringify([80, 95, 100]), "usage.alerts() thresholds === [80,95,100]");
  check(alerts.topup_amounts.length > 0, "usage.alerts() has topup_amounts");
  check(alerts.state.quota_gb === 1.0, "usage.alerts() state.quota_gb === 1.0");
  check(alerts.state.pct_used === 0.0, "usage.alerts() state.pct_used === 0.0 (fresh account)");
  check(Array.isArray(alerts.fired), "usage.alerts() fired is an array");

  // ── webhooks create / list / delete ─────────────────────────────────────
  const url = `https://example.com/videofetch-sdk-test/${Date.now()}`;
  const created = await client.webhooks.create(url, ["download.completed", "quota.exceeded"]);
  info(`webhooks.create(): id=${created.id} events=${JSON.stringify(created.events)} secret_len=${created.secret.length}`);
  check(typeof created.id === "string" && created.id.length > 0, "webhooks.create() returns an id");
  check(created.secret.startsWith("whsec_") && created.secret.length > 22, "webhooks.create() returns full plaintext secret");
  check(!created.secret.includes("*"), "webhooks.create() secret is NOT masked");
  check(created.events.length === 2 && created.events.includes("quota.exceeded"), "webhooks.create() honours the events filter");

  const listed = await client.webhooks.list();
  const found = listed.items.find((w) => w.id === created.id);
  check(!!found, "webhooks.list() contains the created endpoint");
  check(found && found.secret.includes("*") && found.secret !== created.secret, "webhooks.list() secret is masked");

  await client.webhooks.delete(created.id);
  const after = await client.webhooks.list();
  check(!after.items.some((w) => w.id === created.id), "webhooks.delete() removes the endpoint (204)");

  // ── offline signature round-trip ────────────────────────────────────────
  const raw = JSON.stringify({ event: "quota.exceeded", level: "exceeded", threshold_pct: 100 });
  const sig = await computeSignature(raw, created.secret);
  const evt = await constructEvent(raw, sig, created.secret);
  check(evt.event === "quota.exceeded", "constructEvent() verifies a real create() secret");
  await constructEvent(raw, sig, created.secret);
  try {
    await constructEvent(JSON.parse(raw), sig, created.secret);
    check(false, "constructEvent(parsedObject) should throw");
  } catch (e) {
    check(e instanceof SignatureVerificationError, "constructEvent(parsedObject) → SignatureVerificationError");
  }
  check(WEBHOOK_EVENTS.length === 7, "WEBHOOK_EVENTS has 7 events");

  // ── wait() with a small poll interval ────────────────────────────────────
  const job = await client.downloads.create({ url: PROBE_URL, format: "mp3" });
  info(`downloads.create(): id=${job.id} status=${job.status}`);
  check(!!job.id, "downloads.create() returns a queued job");
  let retrieveCalls = 0;
  const origRetrieve = client.downloads.retrieve.bind(client.downloads);
  client.downloads.retrieve = async (id) => { retrieveCalls += 1; return origRetrieve(id); };
  let waitOutcome;
  try {
    const done = await job.wait({ pollIntervalMs: 150, timeoutMs: 1500 });
    waitOutcome = `terminal:${done.status}`;
  } catch (e) {
    // No worker on the dev box → jobs stay queued → wait() times out cleanly.
    waitOutcome = e instanceof VideoFetchErrorLike ? `error:${e.code ?? e.name}` : `error:${e.name}`;
  }
  info(`wait(): outcome=${waitOutcome} retrieveCalls=${retrieveCalls}`);
  check(retrieveCalls >= 1, "wait() polled the API at least once with a small pollIntervalMs");
  check(/terminal:|error:job_timeout|error:VideoFetchError/.test(waitOutcome),
        `wait() resolved or timed out cleanly (${waitOutcome})`);

  // cleanup: cancel the still-queued job so the account stays tidy
  try { await client.downloads.cancel(job.id); info(`downloads.cancel(${job.id}) ok`); }
  catch (e) { info(`downloads.cancel skipped: ${e.name}`); }

  // ── error semantics sanity (RateLimitError shape) ────────────────────────
  const bogus = new VideoFetch({ apiKey: "vf_invalid_key_for_probe", baseUrl: BASE, maxRetries: 0 });
  try {
    await bogus.usage.get();
    check(false, "invalid key should 401");
  } catch (e) {
    check(e.statusCode === 401, `invalid key → HTTP ${e.statusCode} (AuthenticationError)`);
  }
  info(`RateLimitError is exported: ${typeof RateLimitError === "function"}`);

  // ── live 429: fill the account concurrency window, then expect a limit error ─
  const burst = [];
  try {
    for (let i = 0; i < usage.concurrency_limit; i++) {
      burst.push(await client.downloads.create({ url: PROBE_URL, format: "mp3" }));
    }
    info(`created ${burst.length}/${usage.concurrency_limit} in-flight jobs`);
    const err = await client.downloads.create({ url: PROBE_URL, format: "mp3" })
      .then(() => null).catch((e) => e);
    check(err instanceof RateLimitError, "6th concurrent create → RateLimitError (429)");
    if (err instanceof RateLimitError) {
      info(`429: code=${err.code} limit=${err.limit} active=${err.active} scope=${err.scope} retryAfter=${err.retryAfter}`);
      check(["concurrency_limit_exceeded", "queue_limit_exceeded", "platform_at_capacity"].includes(err.code),
            `429 code is a known limit code (${err.code})`);
      check(err.statusCode === 429, "429 statusCode preserved");
      check(typeof err.retryAfter === "number" && err.retryAfter > 0, `Retry-After parsed (${err.retryAfter}s)`);
    }
  } finally {
    for (const j of burst) {
      try { await client.downloads.cancel(j.id); } catch { /* best effort */ }
    }
    info(`cleaned up ${burst.length} burst jobs`);
  }

  console.log(failures === 0 ? `\n== ALL TS INTEGRATION CHECKS PASSED ==\n`
                             : `\n== ${failures} TS INTEGRATION CHECK(S) FAILED ==\n`);
  process.exit(failures === 0 ? 0 : 1);
}

// tiny duck-type helper so we don't import the class just for a name check
const VideoFetchErrorLike = { [Symbol.hasInstance]: (o) => !!o && typeof o === "object" && "code" in o };

main().catch((e) => {
  console.error("\nIntegration crashed:", e);
  process.exit(1);
});

File: python/tests/integration_live.py
"""Live integration test for the official Python SDK (v0.2.0).

用法:
    # 1) 拿一把 key (本地 dev)
    python /tmp/vf_client_provision.py my-email@test.com production > /tmp/vfkey.json
    # 2) 跑
    VF_BASE=http://127.0.0.1:8301 VF_KEY=$(python -c "import json;print(json.load(open('/tmp/vfkey.json'))['api_key'])") \
        python tests/integration_live.py

未设 VF_KEY 时自动跳过 (退出码 0), 便于 CI 无凭据时安全运行。
"""
from __future__ import annotations

import json
import os
import sys

import videofetch
from videofetch import VideoFetch

BASE = os.getenv("VF_BASE", "http://127.0.0.1:8301")
KEY = os.getenv("VF_KEY") or os.getenv("VIDEOFETCH_API_KEY") or ""
VIDEO = "https://www.youtube.com/watch?v=aqz-KE-bpKQ"

if not KEY:
    print("SKIP  no VF_KEY set")
    sys.exit(0)

ok = fail = 0


def chk(name: str, cond: bool, extra: str = "") -> None:
    global ok, fail
    print(("PASS " if cond else "FAIL ") + name + (f"  {extra}" if extra else ""))
    ok += bool(cond)
    fail += (not cond)


def main() -> int:
    c = VideoFetch(api_key=KEY, base_url=BASE)

    # 1) usage (账号级口径 + 告警 + 并发)
    u = c.usage.get()
    chk("usage.quota_gb = 1.0 (free tier)", abs(float(u.quota_gb or 0) - 1.0) < 1e-6, f"quota={u.quota_gb}")
    chk("usage.concurrency_limit present", int(u.concurrency_limit or 0) > 0,
        f"limit={u.concurrency_limit} active={u.active_jobs}")
    chk("usage.alert_level valid", u.alert_level in ("ok", "warning", "critical", "exceeded"),
        f"level={u.alert_level}")
    chk("usage.month non-empty", bool(u.month), f"month={u.month}")

    # 2) alerts
    a = c.usage.alerts()
    chk("alerts.thresholds = [80,95,100]", list(a.thresholds) == [80, 95, 100], f"{a.thresholds}")
    chk("alerts.topup_amounts present", list(a.topup_amounts) == [10, 25, 50, 100], f"{a.topup_amounts}")
    chk("alerts.state has pct/remaining", a.state.pct_used >= 0 and a.state.remaining_gb is not None,
        f"pct={a.state.pct_used} remaining={a.state.remaining_gb} action={a.state.action}")

    # 3) webhooks 用 API key 管理 (v0.2.0 新能力) + 签名
    wh = c.webhooks.create(url="http://127.0.0.1:9/not-listening", events=["download.completed", "quota.warning"])
    chk("webhooks.create → 明文 secret", str(wh.secret).startswith("whsec_"), f"len={len(str(wh.secret))}")
    lst = c.webhooks.list()
    masked = next((w.secret for w in lst.items if w.id == wh.id), "")
    chk("webhooks.list → masked secret", "*" in str(masked) and str(wh.secret) not in str(masked), f"{masked[:14]}…")
    body = b'{"event":"download.completed","id":"dl_test"}'
    import hashlib, hmac
    sig = "sha256=" + hmac.new(str(wh.secret).encode(), body, hashlib.sha256).hexdigest()
    ev = videofetch.construct_event(body, sig, str(wh.secret))
    chk("construct_event 验签通过 (raw body)", ev.get("event") == "download.completed", f"event={ev.get('event')}")
    tampered = False
    try:
        videofetch.construct_event(body + b" ", sig, str(wh.secret))
    except Exception:
        tampered = True
    chk("篡改 body → 抛错", tampered)
    c.webhooks.delete(wh.id)
    chk("webhooks.delete 204", True, "deleted")

    # 4) 真实下载 (mp3 + trim 小任务)
    job = c.downloads.create(url=VIDEO, format="mp3", trim_start=0, trim_end=20)
    chk("downloads.create (扁平 trim 别名)", bool(job.id), f"id={job.id}")
    res = job.wait(timeout=300, poll_interval=3)
    chk("job.wait → completed", res.status == "completed", f"status={res.status} size={res.size_bytes}")
    chk("产物字节 > 0 且带 download_url", (res.size_bytes or 0) > 0 and bool(res.download_url),
        f"size={res.size_bytes} cost={res.cost_usd}")

    # 5) 错误语义
    try:
        VideoFetch(api_key="vf_live_sk_deadbeef", base_url=BASE).usage.get()
        chk("无效 key → AuthenticationError", False)
    except videofetch.AuthenticationError as e:
        chk("无效 key → AuthenticationError", True, f"code={getattr(e, 'code', None)}")

    print(f"\n=== {ok}/{ok + fail} PASSED ===")
    return 1 if fail else 0


if __name__ == "__main__":
    sys.exit(main())
